### ISSUE TITLE
Some doubts regarding values

### ISSUE THREAD

**Issue by gamesterrishi** [sainathadapa/kaggle-freesound-audio-tagging#1]
Hi,

Congratulations on coming 8th on the Kaggle competition. Your model is well thought of and the codebase is very clean and nicely documented. I have a few doubts with respect to the values which have been used:

**07-train-model-mel-and-pca.py**
**pca_train_all = np.load('./data/train_pca_feats.npy')[:, :350]**
Is 350  chosen because we have it has the input dimension for inp2 in our model? Why 350 in model?

**08-generate-predictions.py**
**pca_test = np.load('./data/test_pca_feats.npy')[:, :350]**
Why 350?
**prediction = np.log(np.ones((94, 2)))**
Is 94 the number of test instances?
**for req_mel_len in [263, 363, 463, 563, 663, 763]:**
I understand that these are the number of features to account for 90%,95%,96%,97%,98% variances calculated on train data and transformed on test data. Is my understanding correct? Also, if training data changes we should change these values again, right? And what if we get new holdout test data after we have trained our model - will not having done PCA on that data not affect these values in a wrong way?
**# Geometric average of all the predictions
prediction = prediction / 300.**
Why 300?

**mel_and_pca_model_funcs.py**
**self.req_mel_len = np.random.randint(263, 763)**
Is that the upper and lower bound of our six pca calculations?
** self.req_mel_len_list = [263, 363, 463, 563, 663, 763]**
Same previous question
Above 2 questions similarly for **mel_model_functions.py** too

**summary_feats_funcs.py**
    assert out1.shape[0] == 985
    assert out2.shape[0] == 12
    assert out3.shape[0] == 64
    assert out4.shape[0] == 72
What are these asserts for?

**utils.py**
min_val = -90.5
    max_val = 39.21
Are these values based on the train data? If yes how can one calculate them?

Thanks

**Comment by sainathadapa** [comment]
> 07-train-model-mel-and-pca.py
pca_train_all = np.load('./data/train_pca_feats.npy')[:, :350]
Is 350 chosen because we have it has the input dimension for inp2 in our model? Why 350 in model?
> 08-generate-predictions.py
pca_test = np.load('./data/test_pca_feats.npy')[:, :350]
Why 350?

The first 350 features from `train_pca_feats.npy` account for about 97% of the total variance in the data. I could have just used all the PCA transformed features, but why use the remaining features when they account only 3% of the variance in the data.

> prediction = np.log(np.ones((94, 2)))
Is 94 the number of test instances?

Where did you find the `94` number? Isn't it `prediction = np.log(np.ones((9397, 41)))`? 9397 is the number of test instances.

> for req_mel_len in [263, 363, 463, 563, 663, 763]:
I understand that these are the number of features to account for 90%,95%,96%,97%,98% variances calculated on train data and transformed on test data. Is my understanding correct? Also, if training data changes we should change these values again, right? And what if we get new holdout test data after we have trained our model - will not having done PCA on that data not affect these values in a wrong way?

Audio files in this competition have different lengths. 263 and 763 correspond to 25th and 75th percentile of Mel-Spectrogram feature lengths; nothing to do with the variances. I divided 263-763 range into 5 equal sized bins to get [263, 363, 463, 563, 663, 763].

> Geometric average of all the predictions
prediction = prediction / 300.
Why 300?

- 1 prediction using model from each fold, 10 folds in total
- 1 prediction using each Mel-Spectrogram feature length, 6 such lengths (263, 363, 463, 563, 663, 763)
- 5 random samples each time
So, `10*6*5 = 300`. Refer to the code from line 104 in `08-generate-predictions.py`.

> summary_feats_funcs.py
assert out1.shape[0] == 985
assert out2.shape[0] == 12
assert out3.shape[0] == 64
assert out4.shape[0] == 72
What are these asserts for? 

I wasn't sure about the code, so just added some asserts in the code so that there will be error in case of a bug.


> utils.py
min_val = -90.5
max_val = 39.21
Are these values based on the train data? If yes how can one calculate them?

Load all the Mel-Spectrogram data for train files, and calculate the minimum/maximum values.
